core/include/scipp/core/element/rebin.h

```diff
@@ -13,6 +14,28 @@
 #include "scipp/core/histogram.h"
 #include "scipp/core/transform_common.h"
 
+namespace {
+template <typename NewT, typename OldT>
+void rebinPrimativeType(const NewT &data_new, const OldT &data_old,
+                        scipp::index inew, scipp::index iold,
+                        const double delta, const double owidth) {
+
+  using OldElementT = std::remove_reference_t<decltype(data_old[0])>;
+  using NewElementT = std::remove_reference_t<decltype(data_new[0])>;
+
+  if constexpr (std::is_same_v<bool, OldElementT>) {
+    // Bool -> T is well defined so no cast required
+    data_new[inew] = data_old[iold];
+  } else if constexpr (std::is_same_v<bool, NewElementT>) {
+    // use static_cast to ensure T -> Bool can be done at compile time
+    data_new[inew] = static_cast<NewElementT>(data_old[iold]);
```

Comment:
Why is this required? I don't recall supporting `rebin` with transistions between types?

Also, it looks like you are introducing some bugs here, you are now ignoring a lot of masking values and just default to use that of the last bin contributing to an output bin?

Reply:
On MSVC we are getting warnings since Rebin is transitioning between types, so I'll add a static assert that old / new use the same element type and see what fails to compile

👍 - I'll look into fixing that

Reply:
> On MSVC we are getting warnings since Rebin is transitioning between types

We are not? The output data type should always be the input data type. If other cases are compiled there must be a bug in the type list.

Reply:
I'll have a look, it's either instantiating the templates with extra types we didn't ask for, or there is a bug which is causing some conversions to happen on MSVC / that GCC does not warn about